sdk/src/main/java/io/opentelemetry/sdk/trace/SpanWrapper.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.trace;
+
+import com.google.auto.value.AutoValue;
+import io.opentelemetry.common.AttributeValue;
+import io.opentelemetry.sdk.common.InstrumentationLibraryInfo;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.TraceState;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+@AutoValue
+abstract class SpanWrapper implements SpanData {
+  abstract RecordEventsReadableSpan delegate();
+
+  abstract List<Link> resolvedLinks();
+
+  abstract List<Event> resolvedEvents();
+
+  abstract Map<String, AttributeValue> attributes();
+
+  abstract int totalAttributeCount();
+
+  abstract int totalRecordedEvents();
+
+  abstract Status status();
+
+  static SpanWrapper create(
+      RecordEventsReadableSpan delegate,
+      List<Link> links,
+      List<Event> events,
+      Map<String, AttributeValue> attributes,
+      int totalAttributeCount,
+      int totalRecordedEvents,
+      Status status) {
+    return new AutoValue_SpanWrapper(
+        delegate,
+        links,
+        events,
+        new HashMap<>(attributes),
```

Comment:
Do we need to copy these? Just wrapping in a unmodifiable map should be enough. After the span ends :) before it is needed.

Reply:
good call. I'll make that change and re-benchmark

Reply:
Actually, are you suggesting we do this inside the RERS when the span ends? If we did that, we'd need to have an extra field to hold it, unless we created a version of the AttributesMap that was immutable.

Reply:
AttributeMap is immutable after end.

Reply:
It isn't right now, if we let it escape the RERS. Are you saying we should make it so?

Reply:
Not sure I understand the concern. We call the wrapper using `Collections.unmodifiableMap(attributeMap)`. So nobody can change the attributeMap via SpanData, and nobody can change attributeMap via RERS because hasEnded is true so no changes will happen.

Reply:
Right, if we do the wrapping here, then everything is fine. I thought maybe you were considering doing the wrapping earlier, immediately on the span ending, rather than on the call to `toSpanData`.

Reply:
Also, I think we should wrap the links and events, as well.

Reply:
We can do both later. Let's merge it as it is for the moment. I think this is a good start.